ksql-common/src/main/java/io/confluent/ksql/GenericRow.java

```diff
@@ -15,83 +15,111 @@
 
-  private final List<Object> columns;
+  private final ArrayList<Object> values;
 
   public GenericRow() {
-    columns = new ArrayList<>();
+    this(0);
   }
 
-  public GenericRow(final List<Object> columns) {
-    Objects.requireNonNull(columns);
-    this.columns = columns;
+  public GenericRow(final int initialCapacity) {
+    this.values = new ArrayList<>(initialCapacity);
   }
 
-  public GenericRow(final Object ...columns) {
-    this(Arrays.asList(columns));
+  @VisibleForTesting // Only use from tests
+  public static GenericRow genericRow(final Object... columns) {
+    return new GenericRow().appendAll(Arrays.asList(columns));
+  }
+
+  /**
+   * Ensure the row has enough capacity to hold {@code additionalCapacity} more elements that its
+   * current size.
+   *
+   * <p>Useful to avoid unnecessary array copies when adding multiple elements.
+   *
+   * @param additionalCapacity the number of additional elements
+   */
+  public void ensureAdditionalCapacity(final int additionalCapacity) {
+    values.ensureCapacity(additionalCapacity + values.size());
+  }
+
+  public int size() {
+    return values.size();
+  }
+
+  public Object get(final int index) {
+    return values.get(index);
+  }
+
+  public void set(final int index, final Object value) {
+    values.set(index, value);
+  }
+
+  public GenericRow append(final Object value) {
+    values.add(value);
+    return this;
+  }
+
+  public GenericRow appendAll(final Collection<?> values) {
+    this.values.addAll(values);
+    return this;
+  }
+
+  @JsonProperty("columns")
+  public List<Object> values() {
+    return Collections.unmodifiableList(values);
   }
 
   @Override
   public String toString() {
-    final StringBuilder stringBuilder = new StringBuilder("[ ");
-    int currentIndex = 0;
-    for (int i = 0; i < columns.size(); i++) {
-      final Object obj = columns.get(i);
-      if (obj == null) {
-        stringBuilder.append("null");
-      } else if (obj.getClass().isArray()) {
-        stringBuilder.append(Arrays.toString((Object[]) obj));
-      } else if (obj instanceof String) {
-        stringBuilder.append("'")
-            .append(obj)
-            .append("'");
-      } else {
-        stringBuilder.append(obj);
-      }
-
-      currentIndex++;
-      if (currentIndex < columns.size()) {
-        stringBuilder.append(" | ");
-      }
-    }
-    stringBuilder.append(" ]");
-    return stringBuilder.toString();
+    return values.stream()
+        .map(GenericRow::formatValue)
+        .collect(Collectors.joining(" | ", "[ ", " ]"));
   }
 
   @Override
   public boolean equals(final Object o) {
     if (this == o) {
       return true;
     }
+
     if (o == null || getClass() != o.getClass()) {
       return false;
     }
-    final GenericRow that = (GenericRow) o;
-    if (columns.size() != that.columns.size()) {
-      return false;
-    }
 
-    // For now string matching is used to compare the rows as double comparison will cause issues
```

Comment:
What happened to this concern about double comparison? Are doubles never being compared in tests?

Reply:
So `GenericRow.equals` is only used in tests, so it's only tests that are an issue. 

When I made the change some tests passed because the `GenericRow` was being serialized and deserialized and Jackson is configured to deserializes decimal numbers as `BigDecimal`. So, for example, serializing a generic row with the double value `1.2` would get deserilized as `BigDecimal.valueOf(1.2)`... which would cause `original.equals(deserialised)` to fail.

But I just fixed up the tests to do the right thing.